### PR TITLE
Add possibility to call custom hook scripts on "job done" based on result

### PR DESCRIPTION
**Pull request by okurz**
If a job is done, especially if no label could be found for carry-over,
often more steps are needed for the review of the test result or
providing the information to either external systems or users. As there
can be very custom requirements we should offer a point for
optional configuration to let the instance administrators define
specific actions.

There is already an approach called "auto-review"
https://github.com/os-autoinst/scripts/#auto-review---automatically-detect-known-issues-in-openqa-jobs-label-openqa-jobs-with-ticket-references-and-optionally-retrigger
which offers helpful, external scripts. With this commit it is possible
to call external scripts if defined and based on the selected either
environment variables or config settings. For example config settings
for openqa.opensuse.org enabling the auto-review scripts could look like:

```
[hooks]
job_done_hook_incomplete = /opt/openqa-scripts/openqa-label-known-issues-hook
job_done_hook_failed = /opt/openqa-scripts/openqa-label-known-issues-hook
```

or for openqa.suse.de:

```
[hooks]
job_done_hook_incomplete = env host=openqa.suse.de /opt/openqa-scripts/openqa-label-known-issues-hook
job_done_hook_failed = env host=openqa.suse.de /opt/openqa-scripts/openqa-label-known-issues-hook
```

Related progress issue: https://progress.opensuse.org/issues/77944

## TODO

* [X] Test manually with https://github.com/os-autoinst/scripts/pull/50
* [x] Extend documentation
* [x] Check coverage for newly added code

### CODE REVIEW

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 1]
```diff
@@ -1945,6 +1942,27 @@ sub packages_diff {
...
+corresponding config variable without the C<OPENQA_> prefix can be used in the
+format C<JOB_DONE_HOOK_$RESULT>. The corresponding environment value has
+precedence. The exit code of the externally called script is not evaluated and
+will have no effect.
+
+=cut
+sub done_hook_new_issue {
+    my ($self, $hook) = @_;
+    my $id = $self->id;
+    qx{$hook $id};
```
Comment (Martchus) [inline lib/OpenQA/Schema/Result/Jobs.pm:1963]: I'd rather run this as a Minion job. It should be fairly simple to create a Minion task to run an arbitrary command and enqueue such a job here.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 1 above)
Comment (coolo) [reply lib/OpenQA/Schema/Result/Jobs.pm:1963]: this is a must. Don't run random scripts in a blocking fashion from within the webui - or you'll find yourself in front of 504 pages more often

Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm:1963]: But don't we have enough prefork processes?

Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm:1963]: But yeah, I was thinking of gru jobs as well. Just that I was not sure how to access the app->gru object properly here. I thought maybe it's necessary to to move all to https://github.com/os-autoinst/openQA/blob/master/lib/OpenQA/WebAPI/Controller/API/V1/Job.pm

Comment (coolo) [reply lib/OpenQA/Schema/Result/Jobs.pm:1963]: enough prefork processes for what? We have tons of jobs finishing all the damn time

Comment (Martchus) [reply lib/OpenQA/Schema/Result/Jobs.pm:1963]: You can get the "app" from everywhere via `OpenQA::App->singleton` and via that you can access gru and minion, e.g. to enqueue a job like `OpenQA::App->singleton->enqueue(run_script => [$path])`. (Of course the `run_script` task or something similar still needed to be created first. Just follow the patterns we're already using for other jobs.)

Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm]: solved in https://github.com/os-autoinst/openQA/pull/3603/commits/2eeb04dcd511de973265f9415524ca8cffe9ca00

File: docs/Installing.asciidoc [hunk 2]
```diff
@@ -748,6 +748,35 @@ keys_to_render_as_links=FOO,AUTOYAST
...
+```
+[global]
+JOB_DONE_HOOK_INCOMPLETE = /opt/openqa-scripts/openqa-label-known-issues-hook
+JOB_DONE_HOOK_FAILED = /opt/openqa-scripts/openqa-label-known-issues-hook
+```
+
+or for a host openqa.example.com:
+
+```
+[global]
```
Comment (Martchus) [inline docs/Installing.asciidoc:775]: It would be nicer if these settings had their own section.

File: lib/OpenQA/Task/Job/FinalizeResults.pm [hunk 3]
```diff
@@ -44,6 +45,37 @@ sub _finalize_results {
...
+        if (my $hook = $ENV{'OPENQA_' . $key} // $app->config->{global}->{$key}) {
+            _done_hook_new_issue($openqa_job, $hook);
+        }
+    }
+}
+
+=head2 _done_hook_new_issue
+
+Called if job is done but no issue was carried over. If an environment
+variable corresponding to the found job result is found following the name
```
Comment (Martchus) [inline lib/OpenQA/Task/Job/FinalizeResults.pm:63]:
```suggestion
variable corresponding to the job result is found following the name
```

File: t/10-jobs.t [hunk 4]
```diff
@@ -42,6 +42,20 @@ my $t      = Test::Mojo->new('OpenQA::WebAPI');
...
+sub cover { Devel::Cover::report() if Devel::Cover->can('report') }
+$t->app->minion->on(
+    worker => sub {
+        my ($minion, $worker) = @_;
+        $worker->on(
+            dequeue => sub {
+                my ($worker, $job) = @_;
+                $job->on(cleanup => \&cover) unless $job->info->{notes}{no_cover};
+            });
+    });
```
Comment (Martchus) [inline t/10-jobs.t:55]: It would be better not to C&P this from the GRU test and instead add it to the test utilities.

**Review by Martchus — CHANGES_REQUESTED** [review]

File: t/10-jobs.t (hunk 4 above)
Comment (okurz) [reply t/10-jobs.t]: Yes, I still wanted to do that. Surprise! There were already 20 other duplications :D Fixed in https://github.com/os-autoinst/openQA/pull/3603/commits/7035fbc776f98e92b20ad5c70ebdc64245d161ff

File: docs/Installing.asciidoc (hunk 2 above)
Comment (okurz) [reply docs/Installing.asciidoc]: ok, I selected `[hooks]`

**Comment by okurz** [comment]
updated:
* Implemented suggestions from Martchus
* Bumped timeout in t/10-jobs.t after extending the code called with this PR
* De-duplicated many calls for the minion background process coverage collection fix
* Moved config settions to section `[hooks]` instead of `[global]`
* Removed duplicate documentation comment in lib/OpenQA/Task/Job/FinalizeResults.pm and instead provided template and example in etc/openqa/openqa.ini
* Switched config option format to lower-case

**Comment by okurz** [comment]
Encountered a problem. By moving the carry-over to the background minion job also an AMQP event would not include a ticket ref yet. I am thinking if we should create a separate background minion job for each hook script or move the AMQP event sending into the minion job as well or try to pull out the label carry-over and run it synchronously when ending a job, like before, but in different order, to keep all tasks for the background minion in the same place. I am open for suggestions :)

**Comment by okurz** [comment]
updated:
* Moved label carry-over back out of the minion job to the original place to keep synchronous behaviour. Only the hook scripts call are done additionally in the minion job

**Comment by coolo** [comment]
I would think you might want to extend this (in the future) to mark with certain exit codes of the scripts that carry over shouldn't happen. The whole point of these hooks is flexiblity, no?

As such I would run hooks, carry over, amqp all from within one minion job

**Comment by okurz** [comment]
> I would think you might want to extend this (in the future) to mark with certain exit codes of the scripts that carry over shouldn't happen.

Sorry, I don't understand. The scripts are currently called after carry-over and only if there was no carry-over. This limits the use cases for now of course. My intention is to first focus on the cases where carry-over is not used or not working. Also as it is already possible to disable carry over (per job group at least), this would call hooks in these cases as well when desired. How do you think we could combine the hook scripts with carry-over differently? 

> As such I would run hooks, carry over, amqp all from within one minion job

Yes, I will try that for the future

**Comment by coolo** [comment]
then do the carry over first and pass to the scripts the carried over comment - and they decide what the use case should be (i.e. skip your logic if an argument was passed)

**Comment by okurz** [comment]
> then do the carry over first and pass to the scripts the carried over comment - and they decide what the use case should be (i.e. skip your logic if an argument was passed)

Yes, planned for a follow-up

File: lib/OpenQA/Task/Job/FinalizeResults.pm [hunk 5]
```diff
@@ -44,6 +45,18 @@ sub _finalize_results {
...
+    if (my $hook = $ENV{'OPENQA_' . uc $key} // $app->config->{hooks}->{lc $key}) {
+        _done_hook_new_issue($openqa_job, $hook);
+    }
+}
+
+sub _done_hook_new_issue {
+    my ($openqa_job, $hook) = @_;
+    my $id = $openqa_job->id;
+    qx{$hook $id};
 }
```
Comment (coolo) [inline lib/OpenQA/Task/Job/FinalizeResults.pm:60]: I like the new inline comments by codecov

Comment (okurz) [reply lib/OpenQA/Task/Job/FinalizeResults.pm:60]: this is definitely covered by tests but it seems the coverage data from the minion job is lost even though I tried to apply the fix in t/10-jobs.t . Will check this.

**Comment by okurz** [comment]
updated:
* Changed the minion job to write the hook stdout into minion note and using that in the test when calling the unmocked original hook wrapper method to ensure complete statement coverage. The mocked function calling "orig" would not be seen by Devel::Cover coverage analysis, likely due to the "orig" function not being called from the original scope but only defined temporarily with the same body but not counting as exactly the same function.

**Review by Martchus — APPROVED** [review]

**Review by coolo — COMMENTED** [review]
I only looked at this on a high level, so I won't approve. And if we could trust @Martchus ... :)

**Review by kalikiana — APPROVED** [review]

**Pull request merged**